### PR TITLE
Gene Table Cleanup - data file updates script

### PR DESCRIPTION
**Pull request by rmadupuri**
Script to propagate changes in gene symbols and entrez gene ids in data files.

### What does the script do?
**Step1:**

Replaces the outdated entrez ids to new entrez ids in data files. 

**Step2:**

Updates the hugo symbols in data files based on entrez ids.

1. If entrez id is valid (i.e, it is in either the gene or gene-alias tables)
- If entrez id is only in the gene table, and the hugo symbol in data file does not correspond to the symbol in the gene table for the given entrez id, update the symbol in data file to the symbol from gene table based on the entrez id.
- If entrez id only in the gene-alias table, and the hugo symbol in data file does not correspond to alias symbol for the given entrez id, update the symbol to the alias symbol. If multiple aliases exist for the entrez id, do nothing for now. Have to look into other profiles of the study for mapping these cases.
- If entrez id is in both the gene and gene-alias tables, and the hugo symbol in data file does not correspond to main/alias symbols for the given entrez id, then update the hugo symbol in data file to the symbol from gene table.

2. If entrez id in data file is invalid (i.e, it is not observed in the gene or gene-alias tables)
- If the hugo symbol in the data file is outdated, update it to new hugo symbol.
- If the hugo symbol in data file is `NA`, update the hugo symbol in file to empty cell. `NA` in our vocabulary means Unknown, but in the gene tables it maps to entrez id `7504`. So during the import, there is a chance of incorrectly mapping these records to `7504` due to importer falling back on hugo symbols when entrez id in not known.

**Step3:**

Replaces the outdated hugo symbols to new hugo symbols in data files, if the file has only hugo symbol column.

Every change that is made to data files is logged for reference.

### CODE REVIEW

File: HGNC_Genomic_Files_Migration/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## HGNC Genomic Data Migration Tool
+
+This tool is intended to migrate the changes in entrez ids and hugo symbols based on HGNC in data files. The tool,
```
Comment (yichaoS) [inline HGNC_Genomic_Files_Migration/README.md:3]: "intended to update data files in accordance to <link_to_seed_DB_update>"

File: HGNC_Genomic_Files_Migration/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+1. Replaces the outdated entrez ids/hugo symbols to new entrez ids/hugo symbols. The outdated to new entrez id/hugo symbols mappings are pre-defined based on our analysis and is included in the script folder.
```
Comment (yichaoS) [inline HGNC_Genomic_Files_Migration/README.md:4]: "The ... mappings are actively maintained/updated..".

File: HGNC_Genomic_Files_Migration/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+The tool can be run with the following command:
+```
+python hgnc_data_file_migration.py [-h] -path SOURCE_PATH [-l | -o | -n]
+````
+
+
+**Options**
+```
+-path | --source_path: Path to the data file or directory (of files) that needs to be migrated.
+-l | --stdout_log: Dry-run the script. Preview the changes that will be made to the files.
```
Comment (yichaoS) [inline HGNC_Genomic_Files_Migration/README.md:25]: The "-l" could be skipped here, and instead add this description into overall script description.

File: HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+	#Read the outdated to new hugo symbol mapping file (for cases where the file has only hugo symbol column)
+	outdated_hugo_df = pd.read_csv('outdated_hugo_symbols.txt', sep='\t', header=0, dtype=str)
+	outdated_hugo_dict = dict(zip(outdated_hugo_df['outdated_hugo_symbol'],outdated_hugo_df['new_hugo_symbol']))
+	
+	return main_table_entrez_dict, alias_table_entrez_dict, outdated_entrez_dict, outdated_hugo_dict
+
+#---Check the input type (file/directory) and get the list of files (1 or many)--->
+def check_path(source_path):
+	files_list = []
+	exluded_files_list = ['data_bcr_clinical_data_patient.txt','data_bcr_clinical_data_sample.txt','data_clinical_patient.txt','data_clinical_sample.txt','data_clinical_supp_hypoxia.txt','data_gene_matrix.txt','data_microbiome.txt','data_mutational_signature_confidence.txt','data_mutational_signature_contribution.txt','data_timeline_labtest.txt','data_timeline_procedure.txt','data_timeline_specimen.txt','data_timeline_status.txt','data_timeline_surgery.txt','data_timeline_treatment.txt','data_timeline.txt','data_subtypes.txt','data_fusions.txt']
```
Comment (yichaoS) [inline HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py:40]: `data_fusion` files should be updated as well for outdated symbols?

File: HGNC_Genomic_Files_Migration/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+## HGNC Genomic Data Migration Tool
+
+This tool is intended to migrate the changes in entrez ids and hugo symbols based on HGNC in data files. The tool,
+1. Replaces the outdated entrez ids/hugo symbols to new entrez ids/hugo symbols. The outdated to new entrez id/hugo symbols mappings are pre-defined based on our analysis and is included in the script folder.
+2. Updates the non-matching hugo symbols based on the entrez ids. 
+For example, if the data file contains hugo symbol entrez id pair `MST4	6788`, the hugo symbol `MST4` gets updated to `STK3` since MST4 is not a valid symbol for entrez id 6788.
+
+### Prerequisites
```
Comment (yichaoS) [inline HGNC_Genomic_Files_Migration/README.md:8]: Change to "Requirements"

**Comment by yichaoS** [comment]
1. Are the 2 outdated lists align with the lists in the spread sheet? https://docs.google.com/spreadsheets/d/1KbgGWRV9DsRmQozA_pWHOjTq3Am5OewXWQR8SyVxUlE/edit#gid=0
2. Are the 2 HGNC tables align with the tables in the spreadsheet above?
3. Remove the word "HGNC" from the tables and scripts names - since the updates we added are not all from HGNC, the majority are from NCBI or current DB. 
4. The descriptions of the PR above describes the logic of the script, which is useful and could be included as a separate section in the readme as well.

File: HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+				updated_data += line
+			else:
+				line1 = line.strip('\n').split('\t')
+				entrez = line1[entrez_index]
+				hugo = line1[gene_index]
+				
+				#If entrez id is in outdated list - update the entrez id to new entrez id.
+				if entrez in outdated_entrez_dict:
+					log += entrez+'\t\t---entrez id replaced to---\t\t'+outdated_entrez_dict[entrez]+'\n'
+					line1[entrez_index] = entrez = outdated_entrez_dict[entrez]
```
Comment (yichaoS) [inline HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py:88]: `line1` could be updated to a better variable name.

File: HGNC_Genomic_Files_Migration/README.md (hunk 1 above)
Comment (rmadupuri) [reply HGNC_Genomic_Files_Migration/README.md]: Updated. Pointed the news release link to Gene Table Updates Readme for now. Will have to change the links once the location of the finalized/PR is merged.

File: HGNC_Genomic_Files_Migration/README.md (hunk 2 above)
Comment (rmadupuri) [reply HGNC_Genomic_Files_Migration/README.md]: Done

File: HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py (hunk 4 above)
Comment (rmadupuri) [reply HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py]: Updated the script to handle `data_fusions`. The symbols in `Fusion` column gets updated as well along with `Hugo_Symbol` column.

File: HGNC_Genomic_Files_Migration/README.md (hunk 5 above)
Comment (rmadupuri) [reply HGNC_Genomic_Files_Migration/README.md]: Done

File: HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py (hunk 6 above)
Comment (rmadupuri) [reply HGNC_Genomic_Files_Migration/hgnc_data_file_migration.py]: Done

**Comment by rmadupuri** [comment]
Thank you for the review @yichaoS. 
1. Yes, the two outdated lists align to the spread sheet.
2. Yes, the two HGNC tables align with the tables in the spreadsheet.
3. Good point. Removed the use of word `HGNC`.
4. Added the logic to README.

**Pull request merged**